File: warehouse-api/src/history/history.service.ts
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import { ChangeRecord, ChangeRecordDocument } from '../schemas/change-record.schema';

export interface LogParams {
  userId: string;
  userName: string;
  action: string;
  entity: string;
  entityId?: string;
  description: string;
  businessAction?: string;
  details?: Record<string, any>;
  changes?: Record<string, any>;
  /** Strict per-location tag used by copy/transfer records for exact filtering. */
  locationCode?: string;
}

@Injectable()
export class HistoryService {
  constructor(
    @InjectModel(ChangeRecord.name) private changeModel: Model<ChangeRecordDocument>,
  ) {}

  async clearAll(): Promise<number> {
    const result = await this.changeModel.deleteMany({});
    return result.deletedCount ?? 0;
  }

  async log(params: LogParams) {
    return this.changeModel.create({
      ...params,
      userId: new Types.ObjectId(params.userId),
    });
  }

  private escapeRegex(s: string): string {
    return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }

  async findAll(query: {
    userId?: string;
    action?: string;
    entity?: string;
    keyword?: string;
    businessAction?: string;
    businessActions?: string[];
    locationCode?: string;
    skuCode?: string;
    startDate?: string;
    endDate?: string;
    userName?: string;
    inventoryChangingOnly?: boolean;
    page?: number;
    limit?: number;
  }) {
    const {
      userId, action, entity, keyword, businessAction, businessActions,
      locationCode, skuCode, startDate, endDate, userName,
      inventoryChangingOnly, page = 1, limit = 50,
    } = query;

    const filter: any = {};

    if (userId) filter.userId = new Types.ObjectId(userId);
    if (action) filter.action = action;
    if (entity) filter.entity = entity;
    if (inventoryChangingOnly) filter.entity = 'inventory';

    // businessAction: prefer array OR-match, fall back to single value
    if (businessActions && businessActions.length > 0) {
      filter.businessAction = { $in: businessActions };
    } else if (businessAction) {
      filter.businessAction = businessAction;
    }

    if (userName) filter.userName = { $regex: userName, $options: 'i' };

    const andConditions: any[] = [];
    if (keyword) andConditions.push({ description: { $regex: this.escapeRegex(keyword), $options: 'i' } });

    // SKU + location combined description match (e.g. "ASH008-SY2 @ B2C")
    if (skuCode && locationCode) {
      andConditions.push({
        description: {
          $regex: `${this.escapeRegex(skuCode)} @ ${this.escapeRegex(locationCode)}`,
          $options: 'i',
        },
      });
    } else if (skuCode) {
      andConditions.push({ description: { $regex: this.escapeRegex(skuCode), $options: 'i' } });
    } else if (locationCode) {
      // Prefer strict locationCode field (set on copy/transfer records for exact per-location
      // filtering). Fall back to description regex for old records that lack the field.
      andConditions.push({
        $or: [
          { locationCode: locationCode },
          { locationCode: { $exists: false }, description: { $regex: this.escapeRegex(locationCode), $options: 'i' } },
        ],
      });
    }

    if (andConditions.length > 0) filter.$and = andConditions;

    if (startDate || endDate) {
      filter.createdAt = {};
      if (startDate) filter.createdAt.$gte = new Date(startDate);
      if (endDate) {
        const end = new Date(endDate);
        end.setHours(23, 59, 59, 999);
        filter.createdAt.$lte = end;
      }
    }

    const [records, total] = await Promise.all([
      this.changeModel.find(filter).sort({ createdAt: -1 }).skip((page - 1) * limit).limit(limit).lean(),
      this.changeModel.countDocuments(filter),
    ]);

    return { records, total, page, limit };
  }
}

File: flutter_warehouse/lib/services/history_service.dart
import 'api_service.dart';
import '../models/change_record.dart';

class HistoryService {
  final _api = ApiService.instance.dio;

  Future<Map<String, dynamic>> getAll({
    String? action,
    String? entity,
    String? businessAction,
    String? type,       // IN | OUT | ADJUST | TRANSFER — mapped to businessActions on backend
    String? skuCode,    // filters by "skuCode @ locationCode" in description
    String? keyword,
    String? startDate,
    String? endDate,
    String? userName,
    String? locationCode,
    bool? inventoryChangingOnly,
    int page = 1,
    int limit = 50,
  }) async {
    final response = await _api.get('/audit-logs', queryParameters: {
      if (action != null) 'action': action,
      if (entity != null) 'entity': entity,
      if (businessAction != null) 'businessAction': businessAction,
      if (type != null) 'type': type,
      if (skuCode != null && skuCode.isNotEmpty) 'skuCode': skuCode,
      if (keyword != null && keyword.isNotEmpty) 'keyword': keyword,
      if (startDate != null) 'startDate': startDate,
      if (endDate != null) 'endDate': endDate,
      if (userName != null && userName.isNotEmpty) 'userName': userName,
      if (locationCode != null) 'locationCode': locationCode,
      if (inventoryChangingOnly == true) 'inventoryChangingOnly': 'true',
      'page': page,
      'limit': limit,
    });

    final data = response.data;
    return {
      'records': (data['records'] as List).map((e) => ChangeRecord.fromJson(e)).toList(),
      'total': data['total'],
      'page': data['page'],
    };
  }
}
